exercises/5.3/solution_ws/ros2/src/filter_call/filter_call/filter_call.py

```diff
@@ -0,0 +1,133 @@
+#!/usr/bin/env python
+
+import rclpy
+from rclpy.node import Node
+import py_perception.srv
+from py_perception.srv import FilterCloud
+from sensor_msgs.msg import PointCloud2
+
+class FilterNode(Node):
+    def __init__(self):
+        super().__init__('filter_cloud', allow_undeclared_parameters = False,
+                         automatically_declare_parameters_from_overrides = True)
+        self.client = self.create_client(FilterCloud, 'filter_cloud')
+        while not self.client.wait_for_service(timeout_sec=1.0):
+            self.get_logger().info('service not available, waiting again...')
+
+        self.pcdfilename = self.get_parameter_or('pcdfilename', rclpy.parameter.Parameter('', rclpy.parameter.Parameter.Type.STRING, '')).value
+        self.last_cloud = None
+
+        # PUBLISHERS
+        self.voxel_pub = self.create_publisher(PointCloud2, '/perception_voxelGrid', 1)
+        self.pass_pub = self.create_publisher(PointCloud2, '/perception_passthrough', 1)
+        self.plane_pub = self.create_publisher(PointCloud2, '/perception_planeSegmentation', 1)
+        self.cluster_pub = self.create_publisher(PointCloud2, '/perception_clusterExtraction', 1)
+
+        while rclpy.ok():
+            self.voxel_filter()
+            self.passthrough_filter()
+            self.plane_segmentation()
+            self.cluster_extraction()
```

Comment:
It feels like this while loop should happen outside the class, not in its totally-not-a-constructor init() function.  Or is this just the Pythonic way to do things?  I would have expected a loop outside the function that either calls each of these in order or calls one function that calls all of them.